src/v/storage/segment_reader.cc

```diff
@@ -244,4 +244,86 @@ void segment_reader_handle::operator=(segment_reader_handle&& rhs) noexcept {
     _hook.swap_nodes(rhs._hook);
 }
 
+concat_segment_data_source_impl::concat_segment_data_source_impl(
+  std::vector<ss::lw_shared_ptr<segment>> segments,
+  size_t start_pos,
+  size_t end_pos,
+  ss::io_priority_class priority_class)
+  : _segments{std::move(segments)}
+  , _current_pos{_segments.begin()}
+  , _start_pos{start_pos}
+  , _end_pos{end_pos}
+  , _priority_class{priority_class} {}
+
+ss::future<ss::temporary_buffer<char>> concat_segment_data_source_impl::get() {
+    if (!_current_stream) {
+        _current_stream = co_await next_stream();
+    }
+
+    ss::temporary_buffer<char> buf = co_await _current_stream->read();
+    while (buf.empty() && _current_pos != _segments.end()) {
```

Comment:
Nit: can this `while` be an `if`? Is there ever a case when we need to skip multiple segments? Since we read the segments in order that can only happen if segments can be empty. Is that the case?

Reply:
You're right, it's just a defence against an unlikely case where a segment file happens to be empty. I don't think it will ever happen in the real world, I just added it to make the loop conditions simpler (otherwise we have to get the next handle, then get the buffer, then check if the buffer was empty etc).